File: Difficulty: Hard/Minimize Max Distance to Gas Station/minimize-max-distance-to-gas-station.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    double findSmallestMaxDist(vector<int> &stations, int k) {
        int n = stations.size();
        vector<int> howMany(n-1, 0);
        priority_queue<pair<double, int>> pq;
        for(int i = 0; i < n-1; i++){
            double diff = stations[i+1] - stations[i];
            pq.push({diff, i});
        }
        
        for(int i = 1; i <= k; i++){
            auto topmost = pq.top();
            pq.pop();
            int index = topmost.second;
            howMany[index]++;
            double new_diff = (stations[index+1] - stations[index])/ (double)(howMany[index] + 1);
            pq.push({new_diff, index});
        }
        
        return pq.top().first + 1e-4;
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> stations(n);
        for (int i = 0; i < n; i++) {
            cin >> stations[i];
        }
        int k;
        cin >> k;
        Solution obj;
        cout << fixed << setprecision(2) << obj.findSmallestMaxDist(stations, k)
             << endl;
        cout << "~" << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Hard/Minimize Max Distance to Gas Station/minimize-max-distance-to-gas-station.py
import heapq
class Solution:
    def minMaxDist(self, stations, k):
        pq = []
        n = len(stations)
        howMany = [0] * (n-1)
        if n <= 1:
            return 0.0
        
        for i in range(n-1):
            diff = stations[i+1] - stations[i]
            heapq.heappush(pq, (-diff, i))
            
        for gasStation in range(k):
            sec_len, index = heapq.heappop(pq)
            howMany[index] += 1
            diff = stations[index+1] - stations[index]
            new_sec_len = diff / (howMany[index] + 1)
            heapq.heappush(pq, (-new_sec_len, index))
            
        ans, _ = heapq.heappop(pq)
        return -ans
        